File: sdk/php/src/Resources/WebhooksResource.php
<?php

declare(strict_types=1);

namespace OpenWA\Resources;

use OpenWA\Http\HttpExecutor;

/**
 * Webhooks resource — configure event delivery to external HTTP endpoints.
 *
 * Backed by src/modules/webhook/webhook.controller.ts.
 */
class WebhooksResource
{
    private HttpExecutor $http;

    public function __construct(HttpExecutor $http)
    {
        $this->http = $http;
    }

    /**
     * List webhooks across EVERY session the key can see, not one session's. Requires an
     * OPERATOR-level key.
     *
     * @param array<string,mixed> $query Optional pagination: `limit`, `offset`.
     *
     * @return array<int,array<string,mixed>>
     */
    public function listAll(array $query = []): array
    {
        return $this->http->request('GET', '/api/webhooks', $query) ?? [];
    }

    /**
     * Deliveries that were ATTEMPTED and failed — the diagnostic for a webhook that stopped arriving.
     * Requires an ADMIN-level key. A delivery a smart filter suppressed never reaches this log.
     *
     * @param array<string,mixed> $query Optional filter: `sessionId`, `limit`, `offset`.
     *
     * @return mixed the response has no published schema, so it is returned unshaped
     */
    public function deliveryFailures(array $query = [])
    {
        return $this->http->request('GET', '/api/webhooks/delivery-failures', $query);
    }

    /** @return array<int,array<string,mixed>> */
    public function list(string $sessionId): array
    {
        return $this->http->request('GET', "/api/sessions/{$this->http->encodeSegment($sessionId)}/webhooks") ?? [];
    }

    /** @return array<string,mixed> */
    public function get(string $sessionId, string $id): array
    {
        return $this->http->request('GET', "/api/sessions/{$this->http->encodeSegment($sessionId)}/webhooks/{$this->http->encodeSegment($id)}");
    }

    /**
     * A `secret` in $body signs every delivery as `X-OpenWA-Signature: sha256=<hex>`. The gateway
     * enforces a 16-character minimum on it and answers 400 below that; omit it for unsigned
     * deliveries. Neither `secret` nor `headers` is returned by a read.
     *
     * @param array<string,mixed> $body
     * @return array<string,mixed>
     */
    public function create(string $sessionId, array $body): array
    {
        // headers is a map: an empty PHP array would serialize as a JSON list [] and be rejected by
        // the gateway's object validation. Cast the empty map to stdClass so it encodes as {}.
        if (isset($body['headers']) && $body['headers'] === []) {
            $body['headers'] = new \stdClass();
        }
        return $this->http->request('POST', "/api/sessions/{$this->http->encodeSegment($sessionId)}/webhooks", [], $body);
    }

    /**
     * Partial update: an absent key is left alone. The create route's 16-character minimum on
     * `secret` applies here too, with one exception: `''` is the documented "clear the secret"
     * value and is accepted, as `[]` is for `headers`.
     *
     * @param array<string,mixed> $body
     * @return array<string,mixed>
     */
    public function update(string $sessionId, string $id, array $body): array
    {
        // Same empty-map cast as create(): headers must encode as {} when empty.
        if (isset($body['headers']) && $body['headers'] === []) {
            $body['headers'] = new \stdClass();
        }
        return $this->http->request('PUT', "/api/sessions/{$this->http->encodeSegment($sessionId)}/webhooks/{$this->http->encodeSegment($id)}", [], $body);
    }

    public function delete(string $sessionId, string $id): void
    {
        $this->http->request('DELETE', "/api/sessions/{$this->http->encodeSegment($sessionId)}/webhooks/{$this->http->encodeSegment($id)}");
    }

    /** @return array<string,mixed> */
    public function test(string $sessionId, string $id): array
    {
        return $this->http->request('POST', "/api/sessions/{$this->http->encodeSegment($sessionId)}/webhooks/{$this->http->encodeSegment($id)}/test");
    }
}

File: sdk/java/src/main/java/com/rmyndharis/openwa/resources/WebhooksResource.java
package com.rmyndharis.openwa.resources;

import static com.rmyndharis.openwa.http.Http.encodeSegment;

import com.rmyndharis.openwa.OpenWAClient;
import com.rmyndharis.openwa.http.HttpMethod;
import com.rmyndharis.openwa.model.DeliveryFailureQuery;
import com.rmyndharis.openwa.model.CreateWebhookRequest;
import com.rmyndharis.openwa.model.UpdateWebhookRequest;
import com.rmyndharis.openwa.model.WebhookResponse;
import com.rmyndharis.openwa.model.WebhookTestResult;
import java.util.List;

/** Webhooks resource — configure event delivery to external HTTP endpoints. */
public final class WebhooksResource {
    private final OpenWAClient client;

    public WebhooksResource(OpenWAClient client) {
        this.client = client;
    }

    /**
     * List webhooks across EVERY session the key can see, not one session's. Requires an
     * OPERATOR-level key.
     */
    public List<WebhookResponse> listAll(DeliveryFailureQuery query) {
        return client.requestList(HttpMethod.GET, "/api/webhooks", query, null, WebhookResponse.class);
    }

    /**
     * Deliveries that were ATTEMPTED and failed — the diagnostic for a webhook that stopped arriving.
     * Requires an ADMIN-level key.
     *
     * <p>A delivery a smart filter suppressed never reaches this log. The response has no published
     * schema, so it is returned as a raw JSON tree.
     */
    public Object deliveryFailures(DeliveryFailureQuery query) {
        return client.request(HttpMethod.GET, "/api/webhooks/delivery-failures", query, null, Object.class);
    }

    /** List all webhooks for a session. */
    public List<WebhookResponse> list(String sessionId) {
        return client.requestList(
            HttpMethod.GET, "/api/sessions/" + encodeSegment(sessionId) + "/webhooks", null, null, WebhookResponse.class);
    }

    /** Get a single webhook by id. */
    public WebhookResponse get(String sessionId, String id) {
        return client.request(
            HttpMethod.GET,
            "/api/sessions/" + encodeSegment(sessionId) + "/webhooks/" + encodeSegment(id),
            null,
            null,
            WebhookResponse.class);
    }

    /** Create a new webhook. */
    public WebhookResponse create(String sessionId, CreateWebhookRequest body) {
        return client.request(
            HttpMethod.POST, "/api/sessions/" + encodeSegment(sessionId) + "/webhooks", null, body, WebhookResponse.class);
    }

    /** Update a webhook. */
    public WebhookResponse update(String sessionId, String id, UpdateWebhookRequest body) {
        return client.request(
            HttpMethod.PUT,
            "/api/sessions/" + encodeSegment(sessionId) + "/webhooks/" + encodeSegment(id),
            null,
            body,
            WebhookResponse.class);
    }

    /** Delete a webhook. */
    public void delete(String sessionId, String id) {
        client.requestVoid(
            HttpMethod.DELETE, "/api/sessions/" + encodeSegment(sessionId) + "/webhooks/" + encodeSegment(id), null, null);
    }

    /** Trigger a test dispatch to the webhook URL and report the result. */
    public WebhookTestResult test(String sessionId, String id) {
        return client.request(
            HttpMethod.POST,
            "/api/sessions/" + encodeSegment(sessionId) + "/webhooks/" + encodeSegment(id) + "/test",
            null,
            null,
            WebhookTestResult.class);
    }
}
